lightkurve/io.py

```diff
@@ -0,0 +1,243 @@
+"""Functions for reading light curve data."""
+import logging
+import warnings
+
+from astropy.io import fits
+
+from .lightcurve import KeplerLightCurve
+from .utils import validate_method, LightkurveWarning
+
+log = logging.getLogger(__name__)
+
+__all__ = ['read', 'open', 'read_k2sff']
+
+
+def open(path_or_url, **kwargs):
+    """DEPRECATED. Please use `lk.read()` instead.
+
+    This function has been deprecated because its name collides with Python's
+    built-in `open()` function.
+    """
+    warnings.warn("`lightkurve.open()` is deprecated, please use "
+                  "`lightkurve.read()` instead.",
+                    LightkurveWarning)
+    return read(path_or_url, **kwargs)
+
+
+def read(path_or_url, **kwargs):
+    """Reads any valid Kepler or TESS data file and returns an instance of
+    `~lightkurve.lightcurvefile.LightCurveFile` or
+    `~lightkurve.targetpixelfile.TargetPixelFile`.
+
+    This function will use the `detect_filetype()` function to
+    automatically detect the type of the data product, and return the
+    appropriate object. File types currently supported are::
+
+        * `KeplerTargetPixelFile` (typical suffix "-targ.fits.gz");
+        * `KeplerLightCurveFile` (typical suffix "llc.fits");
+        * `TessTargetPixelFile` (typical suffix "_tp.fits");
+        * `TessLightCurveFile` (typical suffix "_lc.fits").
+
+    Parameters
+    ----------
+    path_or_url : str
+        Path or URL of a FITS file.
+
+    Returns
+    -------
+    data : a subclass of  `~lightkurve.targetpixelfile.TargetPixelFile` or
+        `~lightkurve.lightcurvefile.LightCurveFile`, depending on the detected file type.
+
+    Raises
+    ------
+    ValueError : raised if the data product is not recognized as a Kepler or
+        TESS product.
+
+    Examples
+    --------
+    To read a target pixel file using its path or URL, simply use:
+
+        >>> tpf = read("mytpf.fits")  # doctest: +SKIP
+    """
+    log.debug("Opening {}.".format(path_or_url))
+    # pass header into `detect_filetype()`
+    try:
+        with fits.open(path_or_url) as temp:
+            filetype = detect_filetype(temp)
+            log.debug("Detected filetype: '{}'.".format(filetype))
+    except OSError as e:
+        filetype = None
+        # Raise an explicit FileNotFoundError if file not found
+        if 'No such file' in str(e):
+            raise e
+
+    # Community-provided science products
+    if filetype == "K2SFF":
+        return read_k2sff(path_or_url, **kwargs)
```

Comment:
If you can do this with `read`, do we ever want to expose these helper functions to the user? It might confuse people if they think they have to have a specific read function.

Reply:
I don't mind exposing these functions because (a) they advertise which products we support, and (b) it provides a convenient fall-back option if our file type detection fails.  We can focus on simply advertising `read` in tutorials and documentation however.